google-ads-examples/src/main/java/com/google/ads/googleads/examples/advancedoperations/AddAdCustomizer.java

```diff
@@ -0,0 +1,498 @@
+          feedItemServiceClient.mutateFeedItems(
+              Long.toString(customerId), feedItemOperations);
+
+      List<String> feedItemResourceNames = new ArrayList<>();
+      System.out.printf("Added %d feed items:%n", response.getResultsCount());
+
+      for(MutateFeedItemResult result : response.getResultsList()) {
+        String feedItemResourceName = result.getResourceName();
+        feedItemResourceNames.add(feedItemResourceName);
+        System.out.println(feedItemResourceName);
+      }
+      return feedItemResourceNames;
+    }
+  }
+
+  /**
+   * Helper function to create a FeedItemOperation.
+   *
+   * @param name the value of the Name attribute.
+   * @param price the value of the Price attribute.
+   * @param date the value of the Date attribute.
+   * @param feedResourceName the resource name of the feed.
+   * @param feedAttributes the attributes to be set on the feed.
+   * @return a FeedItemOperation to create a feed item.
+   */
+  private FeedItemOperation createFeedItemOperation(
+      String name, String price, String date, String feedResourceName,
+      Map<String, FeedAttribute> feedAttributes) {
+    FeedItemAttributeValue nameAttributeValue = FeedItemAttributeValue.newBuilder()
+        .setFeedAttributeId(feedAttributes.get("Name").getId())
+        .setStringValue(StringValue.of(name))
+        .build();
+
+    FeedItemAttributeValue priceAttributeValue = FeedItemAttributeValue.newBuilder()
+        .setFeedAttributeId(feedAttributes.get("Price").getId())
+        .setStringValue(StringValue.of(price))
+        .build();
+
+
+    FeedItemAttributeValue dateAttributeValue = FeedItemAttributeValue.newBuilder()
+        .setFeedAttributeId(feedAttributes.get("Date").getId())
+        .setStringValue(StringValue.of(date))
+        .build();
+
+    FeedItem feedItem = FeedItem.newBuilder()
+        .setFeed(StringValue.of(feedResourceName))
+        .addAttributeValues(nameAttributeValue)
+        .addAttributeValues(priceAttributeValue)
+        .addAttributeValues(dateAttributeValue)
+        .build();
+
+    return FeedItemOperation.newBuilder()
+        .setCreate(feedItem)
+        .build();
+  }
+
+  /**
+   * Restricts the feed items to work only with a specific ad group; this prevents the feed items
+   * from being used elsewhere and makes sure they are used only for customizing a specific
+   * ad group.
+   *
+   * @param googleAdsClient the Google Ads API client.
+   * @param customerId the client customer ID.
+   * @param adGroupIds the ad group IDs to bind the feed items to.
+   * @param feedItemResourceNames the resource names of the feed items.
+   */
+  private void createFeedItemTargets(
+      GoogleAdsClient googleAdsClient, long customerId, List<Long> adGroupIds,
+      List<String> feedItemResourceNames) {
+
+    for(int i = 0; i < adGroupIds.size(); i++) {
+      Long adGroupId = adGroupIds.get(i);
+      String feedItemResourceName = feedItemResourceNames.get(i);
```

Comment:
It's a little unclear what's going on here. Are you binding each one of the feedItems to a different adGroup in a 1:1 relationship? If so, why? Either way, it would be helpful to have comments explaining the rationale.

Reply:
Added a comment explaining what's happening here, let me know if you think it makes it clearer.